Named pipe abstraction

angrieralien opened the pull request.

angrieralien commented on ingesters/dirreader/dirreader.go:
copied from internal

angrieralien commented on ingesters/namedpipe/namedpipeingester.go:
ingests lines from a named pipe.

angrieralien commented on internal/journald/doc.go:
don't need `journald` library since we can get the same information from syslog socket using rsyslog. rsyslog tails the syslog socket, filters for sshd events and then writes the entries to a named pipe.

angrieralien commented on processors/auditd/auditd.go:
```diff
@@ -47,19 +40,16 @@ type Auditd struct {
 // TODO: Write documentation about creating a splunk query that shows
 // only events after a user-start.
-func (o *Auditd) Read(ctx context.Context) error {
+func (o *Auditd) Process(ctx context.Context) error {
```
renamed to process for later `pub/sub` abstraction.

angrieralien commented on processors/auditlog/auditlogProcessor.go:
```diff
@@ -0,0 +1,30 @@
+	AuditLogChan chan string
+}
+
+func (a *AuditLogProcessor) Process(ctx context.Context, r io.Reader, readBytes *bytes.Buffer, buf []byte) (int, error) {
```
this func reads in bytes from the namedpipe ingester and pushes log entries to the channel for every newline.

angrieralien commented on processors/sshd/sshdProcessor.go:
```diff
@@ -67,6 +69,52 @@ func SetLogger(l *zap.SugaredLogger) {
+}
+
+func (s *SshdProcessor) Process(ctx context.Context, r io.Reader, currentLog *bytes.Buffer, buf []byte) (int, error) {
+	n, err := r.Read(buf[:cap(buf)])
```
this func reads in bytes from the namedpipe ingester and calls `ProcessEntry` on every newline.

angrieralien commented on processors/sshd/sshdProcessor.go:
```diff
@@ -67,6 +69,52 @@ func SetLogger(l *zap.SugaredLogger) {
+}
+
+func (s *SshdProcessor) ParseSyslogMessage(entry string) syslogMessage {
+	entrySplit := strings.Split(entry, " ")
```
this func extracts the pid and message from the log entry.

angrieralien commented on processors/sshd/sshdProcessor.go:
```diff
@@ -89,7 +137,17 @@ func extraDataWithCA(alg, keySum, certSerial, caData string) (*json.RawMessage,
 }
 
-type processEntryConfig struct {
+func NewSshdProcessor(ctx context.Context, logins chan<- common.RemoteUserLogin, nodeName string, machineID string, eventW *auditevent.EventWriter) *SshdProcessor {
```
Created to pass the `Process` func to the ingester. only variables that change are pid and log entry.

angrieralien commented on rsyslog/config/rsyslog.conf:
rsyslog config that reads from audit.log and syslog socket writing both streams to independent namepipes.

stephen-fox commented on processors/auditlog/auditlogProcessor.go:
```diff
@@ -0,0 +1,30 @@
+package auditlog
```
Please add documentation to the various types and functions in this library.

angrieralien replied:
will do

stephen-fox reviewed (changes requested):
I think this needs to be rebased on `main` once PR #29 is merged. I imagine the `Tail` implementation in this branch will be replaced by whatever we end up using in that PR.

angrieralien commented:
> I think this needs to be rebased on `main` once PR #29 is merged. I imagine the `Tail` implementation in this branch will be replaced by whatever we end up using in that PR.

I wasnt planning on it. cause this tail is specifically for the namedpipe ingester. We dont have to worry about rotating files etc... We could clean it up a bit with the string buffer approach you were talking about if necessary.

angrieralien commented on cmd/run.go:
```diff
@@ -75,91 +79,89 @@ func Run(ctx context.Context, osArgs []string, h *common.Health, optLoggerConfig
-		return fmt.Errorf("failed to get os distro type: %w", err)
+		err := fmt.Errorf("failed to get os distro type: %w", err)
+		logger.Errorf(err.Error())
+		return err
```
will do

angrieralien commented on cmd/run.go:
```diff
@@ -75,91 +79,89 @@ func Run(ctx context.Context, osArgs []string, h *common.Health, optLoggerConfig
-		return fmt.Errorf("failed to get machine id: %w", miderr)
+		err := fmt.Errorf("failed to get machine id: %w", miderr)
+		logger.Errorf(err.Error())
+		return err
```
will do.

angrieralien commented on cmd/run.go:
```diff
@@ -75,91 +79,89 @@ func Run(ctx context.Context, osArgs []string, h *common.Health, optLoggerConfig
-		return fmt.Errorf("failed to get node name: %w", nodenameerr)
+		err := fmt.Errorf("failed to get node name: %w", nodenameerr)
+		logger.Errorf(err.Error())
+		return err
```
will do.

angrieralien commented on cmd/run.go:
```diff
@@ -75,91 +79,89 @@ func Run(ctx context.Context, osArgs []string, h *common.Health, optLoggerConfig
-		return fmt.Errorf("failed to open audit log file: %w", auditfileerr)
+		err := fmt.Errorf("failed to open audit log file: %w", auditfileerr)
+		logger.Errorf(err.Error())
+		return err
```
will do

angrieralien commented on cmd/run.go:
```diff
@@ -75,91 +79,89 @@ func Run(ctx context.Context, osArgs []string, h *common.Health, optLoggerConfig
-			auditLogDirPath, err)
+	eventWriter := auditevent.NewDefaultAuditEventWriter(auf)
+	logins := make(chan common.RemoteUserLogin)
+
```
will do.

angrieralien commented on cmd/run.go:
```diff
@@ -75,91 +79,89 @@ func Run(ctx context.Context, osArgs []string, h *common.Health, optLoggerConfig
+	defer close(auditLogChan)
+
+	auditLogEvents := namedpipe.NamedPipeIngester{
+		FilePath: "/var/log/audit/audit-pipe",
```
we can do that. no problem.

angrieralien commented on cmd/run.go:
```diff
@@ -75,91 +79,89 @@ func Run(ctx context.Context, osArgs []string, h *common.Health, optLoggerConfig
-	}()
+	alp := auditlog.AuditLogProcessor{
+		AuditLogChan: auditLogChan,
+	}
```
will do

angrieralien commented on cmd/run.go:
```diff
@@ -75,91 +79,89 @@ func Run(ctx context.Context, osArgs []string, h *common.Health, optLoggerConfig
+		Audits: auditLogChan,
+		Logins: logins,
+		EventW: eventWriter,
+	}
```
will do

angrieralien commented on cmd/run.go:
```diff
@@ -172,8 +174,11 @@ func Run(ctx context.Context, osArgs []string, h *common.Health, optLoggerConfig
-		return fmt.Errorf("workers finished with error: %w", err)
+		err := fmt.Errorf("workers finished with error: %w", err)
+		logger.Errorf(err.Error())
+		return err
```
will do

JAORMX commented:
Also, if you end up squashing the commits, let's have those be descriptive too.

JAORMX commented on Dockerfile:
```diff
@@ -1,24 +1,25 @@
 COPY --from=builder /go/src/audito-maldito/audito-maldito /usr/bin/audito-maldito
 
-ENTRYPOINT [ "/usr/bin/audito-maldito" ]
+ENTRYPOINT [ "/usr/bin/audito-maldito" ]
```
Given these changes, the journald reader will be broken by default. I suggest you add this to an alternative Dockerfile and have an alternate tag to ship this.

angrieralien replied:
With the namedpipe abstraction we are able to remove the journald ingester since the journald logic is handled by rsyslog. 

I was keeping in there in case there was a future use for journald ingester with audito-maldito. I ended up deleting the jouranld code. We can always pull it back in later if it is necessary.

JAORMX reviewed (changes requested):
This PR is pretty big which has led to us introducing issues in the codebase in the past. Do you have ideas on how we make break this down?

angrieralien commented:
> @angrieralien can you provide a more verbose description in the PR?

**Summary of changes:**
- Creates a concept of ingesters and processors
    - ingesters
        - namedpipe ingeseter: generic ingester that reads from a namedpipe and calls a callback with each newline
        - auditlog ingester: ingests audit.log and passes log entries to the auditLogChannel
        - syslog ingester: ingests journald syslog events and passes log entries to ProcessEntry
        - rocky ingester: ingests rocky sercure logs
    - porcessors
        - auditd processor: same as before
            - removes dirreader.go 
            - updated only for public access
        - sshd processor: same code as processentry.go
            - updated only for public access   
    - updates integration test use the namedpipe
    
**Decision Logic**
-  log ingestion has been solved with other software e.g. rsyslog.
     - rsyslog provides an easy abstraction to read logs from a file and from journald and output those logs to a named pipe.
     - using a tool like rsyslog allows audito-maldito to focus more on extrapolating information from log data rather than collecting log data. 
     - removing the log ingestion code make audito-maldito less complex resulting in easier maintainability.
     - rsyslog can be containerized or run on the host machine
- Incorporates the single responsibility principle by splitting up ingestion logic from the processing logic.
    - Created a ingesters directory where all ingester code lives. Ingester code just focuses on ingesting code
    - Created a processors directory where all process code lives. Processing code just focuses on processing logs.
- Incorporates the open/close principle
    - Splitting up the code by ingesters and processors allows us to extend the code without modifying existing code.
        - new ingesters can be added and use existing processors code.
- Removes fedora dependency
    - now the code is just a go binary without any external dependencies.
- Simplifies supporting new linux distributions
    - add a new ingester struct for the linux distribution with a process function that formats the log data into the desired format for the processors.

angrieralien commented:
> @angrieralien can you provide a more verbose description in the PR?

Ill squash the commits and add the description above.

angrieralien commented:
converting to draft. Going to start integration testing this in some k8s clusters.

angrieralien commented:
final merge in `v0.4.0`
